dll/win32/shlwapi/propbag.cpp

```diff
@@ -995,3 +995,204 @@ SHCreatePropertyBagOnProfileSection(
 
     return pIniPB->QueryInterface(riid, ppvObj);
 }
+
+class CDesktopUpgradePropertyBag : public CBasePropertyBag
+{
+protected:
+    BOOL _AlreadyUpgraded(HKEY hKey);
+    VOID _MarkAsUpgraded(HKEY hkey);
+    HRESULT _ReadFlags(VARIANT *pvari);
+    HRESULT _ReadItemPositions(VARIANT *pvari);
+    IStream* _GetOldDesktopViewStream();
+    IStream* _NewStreamFromOld(IStream *pOldStream);
+
+public:
+    CDesktopUpgradePropertyBag() : CBasePropertyBag(0) { }
+
+    STDMETHODIMP Read(
+        _In_z_ LPCWSTR pszPropName,
+        _Inout_ VARIANT *pvari,
+        _Inout_opt_ IErrorLog *pErrorLog) override;
+
+    STDMETHODIMP Write(_In_z_ LPCWSTR pszPropName, _In_ VARIANT *pvari) override
+    {
+        return E_NOTIMPL;
+    }
+};
+
+VOID CDesktopUpgradePropertyBag::_MarkAsUpgraded(HKEY hkey)
+{
+    DWORD dwValue = TRUE;
+    SHSetValueW(hkey, NULL, L"Upgrade", REG_DWORD, &dwValue, sizeof(dwValue));
+}
+
+BOOL CDesktopUpgradePropertyBag::_AlreadyUpgraded(HKEY hKey)
+{
+    // Written in _MarkAsUpgraded. Just check the existence
+    DWORD dwValue, cbData = sizeof(dwValue);
+    return SHGetValueW(hKey, NULL, L"Upgrade", NULL, &dwValue, &cbData) == ERROR_SUCCESS;
+}
+
+typedef DWORDLONG UNKNOWN_DESKVIEW_FLAGS; // FIXME
+
+HRESULT CDesktopUpgradePropertyBag::_ReadFlags(VARIANT *pvari)
+{
+    UNKNOWN_DESKVIEW_FLAGS Flags;
+    DWORD cbValue = sizeof(Flags);
+    if (SHGetValueA(HKEY_CURRENT_USER,
+                    "Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\DeskView",
+                    "Settings",
+                    NULL,
+                    &Flags,
+                    &cbValue) != ERROR_SUCCESS || cbValue < sizeof(Flags))
+    {
+        return E_FAIL;
+    }
+
+    V_UINT(pvari) = ((UINT)(Flags >> 32)) | 0x220;
+    V_VT(pvari) = VT_UINT;
+    return S_OK;
+}
+
+// FIXME
+typedef struct tagUNKNOWN_STREAM_DATA
+{
+    WORD wMagic;
+    WORD aw[6];
+    WORD wSize;
+} UNKNOWN_STREAM_DATA, *PUNKNOWN_STREAM_DATA;
```

Comment:
Why "Unknown" stream data? Is this name coming from debug symbols?

Reply:
There is no type info abort this.

Reply:
Renamed as `OLD_STREAM_HEADER`.